source/common/upstream/wrsq_scheduler.h

```diff
@@ -0,0 +1,170 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <queue>
+#include <utility>
+#include <vector>
+
+#include "envoy/common/random_generator.h"
+#include "envoy/upstream/scheduler.h"
+
+#include "common/common/assert.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace Upstream {
+
+// Weighted Random Selection Queue (WRSQ) Scheduler
+// ------------------------------------------------
+// This scheduler keeps a queue for each unique weight among all objects inserted and adds the
+// objects to their respective queue based on weight. When performing a pick operation, a queue is
+// selected and an object is pulled. Each queue gets its own selection probability which is weighted
+// as the sum of all weights of objects contained within. Once a queue is picked, you can simply
+// pull from the top and honor the expected selection probability of each object.
+//
+// Adding an object will cause the scheduler to rebuild internal structures on the first pick that
+// follows. This operation will be linear on the number of unique weights among objects inserted.
+// Outside of this case, object picking is logarithmic with the number of unique weights. Adding
+// objects is always constant time.
+//
+// For the case where all object weights are the same, WRSQ behaves identical to vanilla
+// round-robin. If all object weights are different, it behaves identical to weighted random
+// selection.
+//
+// NOTE: This class only supports integral weights and does not allow for the changing of object
+// weights on the fly.
+template <class C> class WRSQScheduler : public Scheduler<C> {
+public:
+  WRSQScheduler(Random::RandomGenerator& random) : random_(random) {}
+
+  std::shared_ptr<C> peekAgain(std::function<double(const C&)>) override {
+    std::shared_ptr<C> picked{pickAndAddInternal()};
+    if (picked != nullptr) {
+      prepick_queue_.emplace(picked);
+    }
+    return picked;
+  }
+
+  std::shared_ptr<C> pickAndAdd(std::function<double(const C&)>) override {
+    // Burn through the pre-pick queue.
+    while (!prepick_queue_.empty()) {
+      auto prepicked_obj = prepick_queue_.front();
+      prepick_queue_.pop();
+      if (!prepicked_obj.expired()) {
+        return std::shared_ptr<C>(prepicked_obj);
+      }
+    }
+
+    return pickAndAddInternal();
+  }
+
+  void add(double weight, std::shared_ptr<C> entry) override {
+    rebuild_cumulative_weights_ = true;
+    queue_map_[weight].emplace(std::move(entry));
+  }
+
+  bool empty() const override { return queue_map_.empty(); }
+
+private:
+  using ObjQueue = std::queue<std::weak_ptr<C>>;
```

Comment:
This may benefit from being a class that automatically sets `rebuild_cumulative_weights_` when it's mutated.